e2e/run_tests.js

```diff
@@ -77,21 +120,128 @@ function writeJsonToFile(jsonObject, filename, dir) {
+    if (!stage && !group) {
+        return allFiles;
+    }
+
+    // Return stage files if no group flag
+    if (stage && !group) {
+        return stageFiles;
+    }
+
+    // Return group files if no stage flag
+    if (!stage && group) {
+        return groupFiles;
+    }
+
+    // Return files if both in stage and group
+    return intersection(stageFiles, groupFiles);
+}
+
+function getSkippedFiles(initialTestFiles, platform, browser, headless) {
+    const platformFiles = grepFiles(grepCommand(`@${platform}`));
+    const browserFiles = grepFiles(grepCommand(`@${browser}`));
+    const headlessFiles = grepFiles(grepCommand(`@${headless ? 'headless' : 'headed'}`));
+
+    const initialSkippedFiles = platformFiles.concat(browserFiles, headlessFiles);
+    const skippedFiles = intersection(initialTestFiles, initialSkippedFiles);
+    const finalTestFiles = without(initialTestFiles, ...skippedFiles);
+
+    // Log which files were skipped
+    if (skippedFiles.length) {
+        console.log(chalk.cyan(`\nSkipped test files due to ${platform}/${browser} (${headless ? 'headless' : 'headed'}):`));
+
+        skippedFiles.forEach((file, index) => {
+            console.log(chalk.cyan(`- [${index + 1}] ${file}`));
+        });
+        console.log('');
+    }
+
+    return {skippedFiles, finalTestFiles};
+}
+
 async function runTests() {
     await fse.remove('results');
     await fse.remove('screenshots');
 
-    const BROWSER = process.env.BROWSER ? process.env.BROWSER : 'chrome';
-    const testDirs = fse.readdirSync('cypress/integration/');
-    let failedTests = 0;
+    const BROWSER = process.env.BROWSER || 'chrome';
+    const HEADLESS = typeof process.env.HEADLESS === 'undefined' ? true : process.env.HEADLESS === 'true';
+
+    const initialTestFiles = getTestFiles().sort((a, b) => a.localeCompare(b));
+    const {finalTestFiles} = getSkippedFiles(initialTestFiles, os.platform(), BROWSER, HEADLESS);
+
+    if (!finalTestFiles.length) {
+        console.log(chalk.red('Nothing to test!'));
+    }
 
     const mochawesomeReportDir = 'results/mochawesome-report';
+    const {invert, group, stage} = argv;
+    let failedTests = 0;
+
+    for (let i = 0; i < finalTestFiles.length; i++) {
+        const testFile = finalTestFiles[i];
+        const testStage = stage ? `Stage: "${stage}" ` : '';
+        const testGroup = group ? `Group: "${group}" ` : '';
+
+        // Log which files were being tested
+        console.log(chalk.magenta.bold(`${invert ? 'All Except --> ' : ''}${testStage}${stage && group ? '| ' : ''}${testGroup}`));
+        console.log(chalk.magenta(`(Testing ${i + 1} of ${finalTestFiles.length})  - `, testFile));
 
-    for (const dir of testDirs) {
         const {totalFailed} = await cypress.run({
-            browser: `${BROWSER}`,
-            headless: 'true',
-            spec: `./cypress/integration/${dir}/**/*`,
+            browser: BROWSER,
+            headless: HEADLESS,
```

Comment:
I'm guessing the possible values for HEADLESS is undefined, headed, headles, true, etc. If it's headed or any other string, would headless parameter recognize the value or would it set it to true if it's a non-empty string?

Reply:
No, values of HEADLESS is either true or false only - https://github.com/mattermost/mattermost-webapp/pull/5107/files#diff-39eb77461c7ea5114f9b1c1a57805effR219.